src/wallet/wallet2.cpp

```diff
@@ -2263,14 +2264,15 @@ bool wallet2::load_keys(const std::string& keys_file_name, const std::string& pa
   // Decrypt the contents
   r = ::serialization::parse_binary(buf, keys_file_data);
   THROW_WALLET_EXCEPTION_IF(!r, error::wallet_internal_error, "internal error: failed to deserialize \"" + keys_file_name + '\"');
-  crypto::chacha8_key key;
-  crypto::generate_chacha8_key(password, key);
+  crypto::chacha_key key;
+  crypto::generate_chacha_key(password, key);
   std::string account_data;
   account_data.resize(keys_file_data.account_data.size());
-  crypto::chacha8(keys_file_data.account_data.data(), keys_file_data.account_data.size(), key, keys_file_data.iv, &account_data[0]);
+  crypto::chacha20(keys_file_data.account_data.data(), keys_file_data.account_data.size(), key, keys_file_data.iv, &account_data[0]);
+  if (json.Parse(account_data.c_str()).HasParseError() || !json.IsObject())
+    crypto::chacha8(keys_file_data.account_data.data(), keys_file_data.account_data.size(), key, keys_file_data.iv, &account_data[0]);
```

Comment:
ohhhh, I see. this is: "try 20, if that doesn't work, fall back to 8"

Reply:
Should we start (covertly) tagging these in some way? An example would be appending a MAC (poly1305?) where the input was `MAC(key, "MONERO-WALLET--FILE-V2" || encrypted data);` The wallet would still just be encrypted bytes (with an implied MAC), so it would contain no identifiable information. And new wallet encryption schemes can be detected easily since the probability of the MAC succeeding is low.

In the current implementation if the MAC failed, it would just jump to the older chacha8 algorithm (can't reject it, unfortunately).